sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1006,7 +1006,6 @@ object EliminateSorts extends Rule[LogicalPlan] {
       case _: Min => true
       case _: Max => true
       case _: Count => true
-      case _: Average => true
```

Comment:
how about `a: Average => !a.child.dataType == FloatType && !a.child.dataType == DoubleType`?

Reply:
oh, pretty smart! I'll update soon.

Reply:
BTW the same argument goes for all central moments, so I think it needs the same treatment

Reply:
Ur, right..., actually, all the aggregates for floating-point values are not commutative, so ISTM we cannot use this opt for these case. I'll update soon.

Reply:
IIUC we cannot apply this opt into `SUM`, too: https://en.wikipedia.org/wiki/Floating-point_arithmetic#Accuracy_problems Can you check again? @cloud-fan @srowen

Reply:
Got it.